### PR TITLE
add(network): Adds a network magic field to `testnet::Parameters` and the config

### LINKED ISSUE #7845: Tracking: Add parameters to `Network::Testnet` variant
## Motivation

Zebra currently hard-codes some consensus parameters that may need to be changed for testing.

## Scope

To be discussed in https://github.com/ZcashFoundation/zebra/discussions/7967.

#### Preparatory Refactors:
- [x] #8325
- [x] #8326
- [x] #7968

After these changes are merged, we should decide which parameters we want to add and create more issues like #7970 for each of them.

#### Adding Parameters:
- [x] #7970
  - [x] #7840
- [x] #8361
- [x] #8362
- [x] #8327
- [x] #8412
- [ ] Add a `default_port` field to `NetworkParameters` (needed for custom Testnets but not Regtest)
- [x] #8364 
- [x] #8365
- [x] #8366
- [x] #8367

### Possible Design

- Turn the `Testnet` variant into:
 ```rust
Testnet(Arc<NetworkParameters { 
    cache_name,
    genesis_hash,
    network_id,
    default_port,
    activation_heights,
    ..
>)
```
- Consider adding fields for:
  - `slow_start_interval` (nSubsidySlowStartInterval)
  - `target_difficulty_limit` (powLimit)
  - `key_prefixes`
  - `debug_mandatory_checkpoint_height: Option<Height>`
  - `checkpoints: Path`
  - `pow_avg_window` (update logic to use `min(num_blocks_in_cache, 17)`)
- Add a `new_testnet()` method or `DEFAULT_TESTNET` constant with the existing testnet defaults, to replace literal `Testnet` variant.
- Replace matching against `Testnet` with an `is_mainnet()` method. (This is only required when we're matching against `Testnet`, we can change the `Mainnet` matches gradually.)

### PR DESCRIPTION
**Pull request by arya2**
## Motivation

We want to allow for configuring the network magic on custom testnets.

Part of #7845.
Depends-On: #8518.

#### Check before marking the PR as ready for review:
  - [x] Will the PR name make sense to users?
  - [x] Does the PR have a priority label?
  - [x] Have you added or updated tests?
  - [x] Is the documentation up to date?

##### For significant changes:
  - [x] Is there a summary in the CHANGELOG?
  - [x] Can these changes be split into multiple PRs?

_If a checkbox isn't relevant to the PR, mark it as done._

## Solution

- Moves network `Magic` type from zebra-network to zebra-chain
- Adds a `network_magic` field on `testnet::Parameters` and uses it in the network codec
- Reads a network magic from the config

### Testing

- Adds a stored test config
- Updates a `Magic` debug impl test
- Adds a test to check that the codec returns an error when messages have unexpected network magics

We could also add a test for the constraints in `testnet::ParametersBuilder::with_network_magic()` (can't be the network magic of a default network) if we want.

## Review

Anyone can review.

### Reviewer Checklist

Check before approving the PR:
  - [ ] Does the PR scope match the ticket?
  - [ ] Are there enough tests to make sure it works? Do the tests cover the PR motivation?
  - [ ] Are all the PR blockers dealt with?
        PR blockers can be dealt with in new tickets or PRs.

_And check the PR Author checklist is complete._

## Follow Up Work

Accept peer IPs in the config as initial peers.

### CODE REVIEW

File: zebra-chain/src/parameters/network/testnet.rs [hunk 1]
```diff
@@ -139,6 +144,20 @@ impl ParametersBuilder {
         self
     }
 
+    /// Sets the network name to be used in the [`Parameters`] being built.
+    pub fn with_network_magic(mut self, network_magic: Magic) -> Self {
+        assert!(
+            [magics::MAINNET, magics::TESTNET, magics::REGTEST]
```
Comment (arya2) [inline zebra-chain/src/parameters/network/testnet.rs]: The testnet magic should be allowed (it's the default value anyways if this method isn't called before `.finish()`), but maybe only on compatible testnets? (currently where the activation heights are the same as the default testnet, as we add more fields, the target difficulty, equihash parameters, and other parameters may also be relevant).

```suggestion
            [magics::MAINNET, magics::REGTEST]
```

File: zebra-chain/src/parameters/constants.rs [hunk 2]
```diff
@@ -15,3 +15,15 @@ pub const SLOW_START_INTERVAL: Height = Height(20_000);
 ///
 /// This calculation is exact, because `SLOW_START_INTERVAL` is divisible by 2.
 pub const SLOW_START_SHIFT: Height = Height(SLOW_START_INTERVAL.0 / 2);
+
+/// Magic numbers used to identify different Zcash networks.
```
Comment (oxarbitrage) [inline zebra-chain/src/parameters/constants.rs:19]: Should we add a link pointing to where did we got these ?

Comment (arya2) [reply zebra-chain/src/parameters/constants.rs:19]: I'm not sure where these are specified (I didn't find it in the protocol specification), but I added a link to where I found it in the zcashd code in https://github.com/ZcashFoundation/zebra/pull/8524/commits/f2bef3a8bfbb8f15f97178eaad04ef66bf7e5469.

**Review by oxarbitrage — APPROVED** [review]
Looks good. Thanks.

**Pull request merged**